### ISSUE TITLE
Getting the width of a street section from a PML file

### ISSUE THREAD

**Issue by vvoovv** [prochitecture/blosm#67]
I'd like to define widths of street sections and other attributes in a PML file.

A PML style is chosen based on the category of an OSM way, the number of lanes, etc.

If possible, an instance of the class `StreetSection` should be created earlier in the code.  A renderer exposes a method `getStyle` which accepts `tags` as a parameter and returns a style defined in PML. The attributes of the instance of the class `StreetSection` (width, the number lanes) are set in accordance with the values in the PML style.

How easy would it be to refactor the existing code?

**Comment by polarkernel** [comment]
> How easy would it be to refactor the existing code?

I don't think this is too complicated. The instances of `StreetSection` are constructed at the end of the `StreetGenerator` processes. Before that, an internal and complicated class `WaySection`, defined in _way/way_section.py_ is used. Its constructor creates it from the section network in the `StreetGenerator`'s [createWaySections()](https://github.com/prochitecture/blosm/blob/ac5e071b2ef01db700c6b6f8914e28472ba2ff44/action/generate_streets.py#L453) function. This is also where the way widths are set, based on the `wayCategoryProps` in _way_properties.py_, by calling `estimateWayWidths()`. The lane patterns for the turn lanes are also set here.

In [createWaySections()](https://github.com/prochitecture/blosm/blob/ac5e071b2ef01db700c6b6f8914e28472ba2ff44/action/generate_streets.py#L453), the method `getStyle` could be called, and their returned parameters could replace the default ones in `WaySection`.

**Comment by vvoovv** [comment]
What is the difference between the classes `NetSection`, `WaySection `and `StreetSection`?

**Comment by polarkernel** [comment]
> What is the difference between the classes NetSection, WaySection and StreetSection?

Well, the `NetSection` class holds the edges in the instances (`networkGraph`, `sectionNetwork`) of the `WayNetwork` class. It is only related to the networks.

The `WaySection` class is the working horse of way processing. Unlike the instances of `NetSection`, there is only one (directed) instance per way section. Its polyline is represented by an instance of the `PolyLine` class, which provides powerful methods for parametrization, the calculation of intersections, trimming, distances, projections, offset lines, and so on. `WaySection` is also used to keep track of lanes and their widths.

Finally, the class `StreetSection` (formerly `TrimmedWaySection`) is used as the output class. Classes internal to street generation (for instance `PolyLine`) are omitted, so that only Python classes or Blender classes are used there.

**Comment by vvoovv** [comment]
What would be the obstacles to merge `WaySection` and `StreetSection` into a single class?

**Comment by polarkernel** [comment]
> What would be the obstacles to merge WaySection and StreetSection into a single class?

The main differences I see are

- Not all sections in the _WaySections_ dictionary are valid. Some have been invalidated because they were too short or were covered by clusters or for other reasons. If we used _WaySections_ instead of _StreetSections_, the renderer would have to test for _isValid_ on each instance.
- The polyline of the _WaySection_ keeps the original length of the section as given by the OSM data. The shorter trimmed centerline, e.g. truncated by intersections, must be calculated and stored anyway.
- The keys of the _WaySection_ dictionary are different from the keys of the _StreetSection_s dictionary, but I can't say yet if this would be a problem.
- The lane data in _StreetSections_ is different in details from that in _WaySections_. _StreetSections_  would require its own attributes.

**Comment by vvoovv** [comment]
I suggest gradually merging `WaySection` and `StreetSection`. But first I'd like to integrate reading PML attributes in the method `createWaySections()` of the class `StreetGenerator`.

**Comment by polarkernel** [comment]
> I suggest gradually merging WaySection and StreetSection.

OK. Just out of curiosity: What's the benefit of that?

**Comment by vvoovv** [comment]
> OK. Just out of curiosity: What's the benefit of that?

No need to create an additional structure.
